### PR TITLE
Snow: fall when constructed over snow

### PR DESCRIPTION
**Pull request by lupoDharkael**
If you add snow over a node of snow it floats in the air and it looks weird:
capture-13 35 28

It now falls (replicates the same behavior as when falling from other heights because there are no nodes below).

### CODE REVIEW

File: mods/default/nodes.lua [hunk 1]
```diff
@@ -530,9 +530,12 @@ minetest.register_node("default:snow", {
 	}),
 
 	on_construct = function(pos)
-		pos.y = pos.y - 1
-		if minetest.get_node(pos).name == "default:dirt_with_grass" then
-			minetest.set_node(pos, {name = "default:dirt_with_snow"})
+		local belowPos = pos
+		belowPos.y = belowPos.y -1
```
Comment (SmallJoker) [inline mods/default/nodes.lua]: Don't.
Either use `table.copy()` or `vector.new` to drop the reference.

Comment (lupoDharkael) [reply mods/default/nodes.lua]: fixed! (and thanks, I started learning Lua yesterday)

**Comment by paramat** [comment]
:-1: There's no point in altering this behaviour because this is just how a voxel game works, nodes will 'float' over other partial nodes because the node below is occupied. Anything built over a slab does the same but we're not going to add code to make slabs fall. Any node built over another node, remove the lower node and the upper node floats.

How did you get the situation in the screenshot? Snow is buildable_to so a placed snow will remove the snow below. If snow is above a node and you remove the lower it falls. Did you place the upper snow by placing it against the tree? If so that's a very rare case so even less reason to add code.

**Comment by lupoDharkael** [comment]
Yes, I placed the snow against the tree. Ad I feel this is an inconsistent behaviour because if you place snow on top of a snow node it floats, but if you do the same just +1 in the **y** axis it falls and disappears.
It may seem to be a very little detail but that's what differentiates a polished game from the others.
But I understand if this is considered a corner case which is not worth correcting.

**Comment by C1ffisme** [comment]
@paramat 
> There's no point in altering this behaviour because this is just how a voxel game works, nodes will 'float' over other partial nodes because the node below is occupied. Anything built over a slab does the same but we're not going to add code for that situation. Any node built over another node, remove the lower node and the upper node floats.

Except slabs aren't `buildable_to` so if sand sits on top of them, it makes a little bit of sense. (I think in Minecraft, the sand actually pops into a dropped item.) Snow is `buildable_to` so it makes a little bit more sense for the snow above it to fall and replace the old snow.

**Comment by paramat** [comment]
Yes, however i still don't think it's worth doing.

**Comment by tenplus1** [comment]
how about adding snow levels to the game so that we could have gradually inclined snow fields :)

**Comment by SmallJoker** [comment]
@tenplus1 There used to be snow levels, which were most likely removed with the buggy weather features. It was a nice feature, but requires core and MTG changes.
Core: item entity, checking for levelled nodes
MTG: `on_place` and `on_dig` code for snow placing and digging.

**Comment by paramat** [comment]
MTG dev opinions needed.

File: mods/default/nodes.lua [hunk 2]
```diff
@@ -530,9 +530,12 @@ minetest.register_node("default:snow", {
 	}),
 
 	on_construct = function(pos)
-		pos.y = pos.y - 1
-		if minetest.get_node(pos).name == "default:dirt_with_grass" then
-			minetest.set_node(pos, {name = "default:dirt_with_snow"})
+		local belowPos = vector.new(pos)
```
Comment (paramat) [inline mods/default/nodes.lua]: `belowpos` or `below_pos`
No need to use 'vector':
`local below_pos = pos`

File: mods/default/nodes.lua [hunk 3]
```diff
@@ -530,9 +530,12 @@ minetest.register_node("default:snow", {
...
+		belowPos.y = belowPos.y -1
```
Comment (paramat) [inline mods/default/nodes.lua]: Space after '-'

File: mods/default/nodes.lua [hunk 4]
```diff
@@ -530,9 +530,12 @@ minetest.register_node("default:snow", {
...
+		if minetest.get_node(belowPos).name == "default:dirt_with_grass" then
+			minetest.set_node(belowPos, {name = "default:dirt_with_snow"})
+		elseif minetest.get_node(belowPos).name == "default:snow"  then
+			minetest.spawn_falling_node(pos)
```
Comment (paramat) [inline mods/default/nodes.lua]: `minetest.get_node()` is intensive, so store `minetest.get_node(belowPos).name` in a variable, then use that variable twice.

**Comment by lupoDharkael** [comment]
maybe @SmallJoker is right about this. Feel free to close this if you want, I agree this is not the best implementation and it could be achieved with a better support for levels in the future.

**Comment by CasimirKaPazi** [comment]
Voxelgarden has this.

[Snow is a falling node](https://github.com/CasimirKaPazi/Voxelgarden/blob/master/mods/default/nodes.lua#L1001)
[Falling increases levels of leveled nodes](https://github.com/CasimirKaPazi/Voxelgarden/blob/master/mods/physics/falling.lua#L66)

**Comment by DS-Minetest** [comment]
And [the snow mod](https://github.com/Splizard/minetest-mod-snow/).

**Comment by paramat** [comment]
Ok closing then as the author is unsure, but feel free to discuss here.

**Pull request closed without merging**